### PR TITLE
[CR] Units wrapper class for volume 

### PR DESCRIPTION
**Pull request by BevapDin**
Adds a wrapper class `quantity<T,U>`, which contains a scalar value of type `T` (e.g. `int`) and can be used instead of a plain `T`. `U` is a tag type that defines the type of the quantity and implicitly its unit.

The `quantity` type can be used similar to numbers, it allows:
- addition and subtraction of other `quantity` values having the same tag type `U`, returning a `quantity`,
- multiplication and division with a scalar (dimensionless) value, returning a `quantity`,
- division with another `quantity` value having the same tag type `U`, returning a scalar value,
- the usual comparison of quantities having the same tag type `U`.

Other operations (e.g. adding a dimensionless value) are not supported on purpose. `quantity` values with different tag types `U` can not be combined in any way.

This allows:
- Strict type checking at compile time. Prevents combining things in a nonsensical way like adding a volume to a mass (they would have different `U` types). Or comparing a volume with a mass.
- Clear, implicit documentation (`get_storage` returns a volume, not some arbitrary integer).
- Better handling of different units: One has to convert any dimensionless value into a `quantity` object by calling explicit, named functions that indicate clearly what the expected input unit is. Same for output.

A typedef defines `volume` as `quantity<int, volume_in_milliter_tag>`, which is now used instead of the plain `int`. Internally it uses milliliter as unit and `int` as its type, but that can be changed easily.

``` C++
// So what unit is?
int some_volume = 3;
// what does 22 mean in real-life values?
if( some_item.volume() < 22 ) {
    ... // give some bonus
}

// new:
units::volume some_volume = units::from_milliliter( 750 );
// or (equivalent) with user-defined literals
using namespace units::literals;
units::volume some_other_volume = 750_ml;
if( some_item.volume() < units::from_liter( 2 ) ) {
    // bonus when item is smaller than 2 liter.
}

// Accidentally called `weight` instead of `volume`:
bool can_pick = u.volume_carried() + item.weight() < u.volume_capacity();
// Would compile just fine, but will now raise an error as one can not add units::volume
// (from volume_carried) and an int (from weight).

// or add a conversion for your preferred unit:
units::volume jug_size = from_gallon(10);
units::volume storage = from_imperial_ounce(10);
```

One can use `from_milliliter` / `from_liter` and the user defined literal `_ms` to convert raw numbers to the new type.

There is also a `legacy_volume_factor` (represents 250 ml) that represents one unit of the current volume system. One can multiply raw numbers of the current system to get the new volume type or divide with it. It's only used to keep old code (and I/O) working without any changes. It should be replaced by an explicit value, which can often be used to reduce the whole term (e.g. `10*vol/legacy_volume_factor` => `vol/25_ml`).

---

The changes in the final diff are mostly:
- Change variable/member/parameter type from `int` to `units::volume` - should not affect the behavior.
- Switch to new types in calculations: `i.volume() / 2` -> `i.volume() / 500_ml`
- Code that still needs raw numbers (e.g. for invoking the RNG, or for I/O) usually gets them by dividing the volume by `units::legacy_volume_factor`. Or by explicitly stating the unit via `from_milliliter` / `to_milliliter`.

Many places that would use `legacy_volume_factor` have been changed to use `from_milliliter` or user defined literals:

``` C++
// old:
ret -= tmp.volume() / 4;
// new:
ret -= tmp.volume() / units::from_liter( 1 );
// or:
ret -= tmp.volume() / 1000_ml;
```

---

Saves and JSON data are currently unaffected. All volumes (e.g. from a folded player-made vehicle item, or artifact) are converted to/from the current system (using `legacy_volume_factor`) when loading/saving data. JSON data is interpreted in the current system and converted to the new system when it's loaded.

---

Finally one big change: volume (`itype::volume`) is now per **single** charge. The value from JSON is interpreted as before (per stack) and is converted when it's loaded.

All items now have an exact volume (measured in milliliters). Most display code has been changed to show milliliters, which looks kind of ugly when the numbers are large ([CR]).

All items have at least 1 ml volume, there are no items with 0 volume. If you only wear a skirt (storage: 250 ml), you can carry up to 250 USB drives, or 250 cash cards, or any combination of at most 250 of those items.

---

To be done next time [CR]:
- Allow volume in JSON in milliliter precision (new member "volume_in_ml" to keep compatibility with old system or parse string with unit, e.g. `"volume": "250 ml"`).
- Store values using milliliter (artifact item volume, folded vehicle volume).
- Consistent display format: in liters as decimal number / in milliliter / user defined?

### CODE REVIEW

File: src/iexamine.cpp [hunk 1]
```diff
@@ -1836,7 +1837,7 @@ void iexamine::kiln_full(player &, const tripoint &examp)
     }
 
     item result( "charcoal", calendar::turn.get_turn() );
-    result.charges = total_volume * char_type->ammo->def_charges / char_type->volume;
+    result.charges = total_volume / char_type->volume;
```
Comment (Coolthulhu) [inline src/iexamine.cpp]: How does this keep the old values?

Comment (BevapDin) [inline src/iexamine.cpp]: I suspect this is a pre-existing bug, it should have used `stack_size` in the first place. Or should have explained why it uses `def_charges`.

Currently charcoal has `def_charges` 50 and `stack_size` 10. A stack of 20 unfinished charcoal (volume 20), created from 5 logs (volume 200) will transform into 1000 charcoal (volume 100). The last transformation (volume 20 -\> volume 100) seems wrong.

This should probably just call `result.charges_per_volume( total_volume )`, which does the same as the new code here. The new logic here preserves the volume of the item.

Comment (Coolthulhu) [inline src/iexamine.cpp]: OK, the current values are too high.

But there is still a problem here after those changes: volume per unit is `volume / stack_size`, but in the version here it is treated just as `volume`.

Comment (BevapDin) [inline src/iexamine.cpp]: `itype::volume` is now per single charge, so `total_volume / volume_per_single_charge` gives the number of charges that make up that volume. That's what `charges_per_volume` calculates.

**Comment by mugling** [comment]
> One can use from_milliliter / from_liter and the user defined literal _ms to convert raw numbers to the new type.

Typo? Did you mean `_ml` - this is significant as C++14 includes an `ms` literal for milliseconds.

File: src/bionics.cpp [hunk 2]
```diff
@@ -389,8 +389,9 @@ bool player::activate_bionic( int b, bool eff_only )
         bool extracted = false;
         for( auto it = g->m.i_at(pos()).begin();
              it != g->m.i_at(pos()).end(); ++it) {
+            static const auto volume_per_water_charge = units::from_milliliter( 500 );
```
Comment (mugling) [inline src/bionics.cpp]: Why not 250?

File: src/generic_factory.h [hunk 3]
```diff
@@ -985,4 +986,65 @@ typename std::enable_if<std::is_constructible<T, std::string>::value, bool>::typ
...
 
+inline bool assign( JsonObject &jo, const std::string &name, units::volume &val,
+                    bool strict = false,
+                    const units::volume lo = units::volume( std::numeric_limits<units::volume::value_type>::min(),
+                            units::volume::unit_type{} ),
+                    const units::volume hi = units::volume( std::numeric_limits<units::volume::value_type>::max(),
+                            units::volume::unit_type{} ) )
+{
+    // Currently JSON data contains volume in 250ml units.
+    // TODO: change JSON to contain milliliter values.
```
Comment (mugling) [inline src/generic_factory.h]: prefer doxygen form of todo

Comment (mugling) [inline src/generic_factory.h]: once we have 100% JSON linter coverage that transition could be easily performed via `jq`

File: src/item.cpp [hunk 4]
```diff
@@ -1371,7 +1347,7 @@ std::string item::info( bool showtext, std::vector<iteminfo> &info ) const
             temp1 << _( "<good>preserves spoiling</good>, " );
         }
 
-        temp1 << string_format( _( "can store <info>%.2f liters</info>." ), c.contains / 4.0 );
+        temp1 << string_format( _( "can store <info>%.2f liters</info>." ), c.contains / 1000.0_ml );
```
Comment (mugling) [inline src/item.cpp]: why not `1000_ml`?

File: src/item.cpp [hunk 5]
```diff
@@ -2521,28 +2499,20 @@ int item::weight( bool include_contents ) const
...
     switch( corpse_size ) {
-        case MS_TINY:    return    3;
-        case MS_SMALL:   return  120;
-        case MS_MEDIUM:  return  250;
-        case MS_LARGE:   return  370;
-        case MS_HUGE:    return 3500;
+        case MS_TINY:    return    750_ml;
+        case MS_SMALL:   return  30000_ml;
+        case MS_MEDIUM:  return  62500_ml;
+        case MS_LARGE:   return  92500_ml;
```
Comment (mugling) [inline src/item.cpp]: Would be reasonable to update these to 'round' numbers

File: src/item.cpp (hunk 4 above)
Comment (BevapDin) [inline src/item.cpp]: The same reason why it was `4.0` and not `4` in the old code: the result should be a `double`. `1000_ml` has `int` as value type and `1000.0_ml` has `double` as value type, thereby keeping the existing behavior.

File: src/bionics.cpp (hunk 2 above)
Comment (BevapDin) [inline src/bionics.cpp]: Previous logic is `it->volume() / 2`, now it must be `it->volume() / 500` because 2 in current system represents 500 ml. I have no idea why this exact value had been chosen.

File: src/generic_factory.h (hunk 3 above)
Comment (BevapDin) [inline src/generic_factory.h]: Do you mean to use `///` instead of plain `//` so Doxygen extracts this?

> once we have 100% JSON linter coverage that transition could be easily performed via  jq

Yes, but external mods will *silently* break (items will have a wrong volume). I'd prefer to have a new member and falling back to the current system if the new member is not present. Anyway, it's not part of this PR.

Comment (mugling) [inline src/generic_factory.h]: > Do you mean to use /// instead of plain // so Doxygen extracts this?

Or even better `@todo`

> Yes, but external mods will silently break (items will have a wrong volume).

We can update mods that are distributed as part of the project in the same manner. External mods are frequently subject to arbitrary breakage in any case. Having two competing systems would be a significant problem - this PR as it stands is fine but the JSON switchover should be 100%

Comment (BevapDin) [inline src/generic_factory.h]: >> Do you mean to use /// instead of plain // so Doxygen extracts this?

> Or even better  @todo

I'll make another PR that converts all existing `TODO`s to that format. The plain, non-Doxygen format is currently in the majority.

File: src/item.cpp (hunk 5 above)
Comment (BevapDin) [inline src/item.cpp]: I tried to produce as few as possible visible changes. Changing the numbers here requires the whole balancing thing to be done.

File: src/generic_factory.h (hunk 3 above)
Comment (mugling) [inline src/generic_factory.h]: Fair enough

**Comment by mugling** [comment]
I can't see any problems with this. Obviously this PR is the start of a series - can you outline the next steps you want to take?

**Comment by BevapDin** [comment]
> Obviously this PR is the start of a series - can you outline the next steps you want to take?

Changing the JSON data is a separate step as it would break compatibility. Adding a new member (something like "volume_in_ml" or similar) and keeping the meaning of the existing "volume" member as fallback would avoid that breakage, but would require more code. (one PR)

Changing the values in the saves is simple as the savegame version can be used to detect what unit the stored data represents (old system or new system). (another PR)

I plan to convert weight and possibly temperature to this system. Theoretically one could also convert radiation, hunger, thirst, fatigue, item charges etc. - but that might be excessive. (several more PRs)
